refactor: unify XBlock naming: item [BD-13]

tecoholic opened the pull request:
## Description

Currently, there are various terms used to refer to XBlocks: xblocks (or blocks), descriptors, modules, or items. This makes the code a lot less approachable and understandable. Since we are getting rid of a lot of tech debt it is a good time to do a quick update of the terminology in the edx-platform to make it consistent.

In this PR:

Convert references using `item` to refer to XBlocks to use `block` (e.g. in ItemFactory). 


## Supporting information

This ticket was created from the work started with https://github.com/openedx/edx-platform/pull/31113

## Testing instructions

There is nothing specific to be tested, as we are not adding or changing any functionality. However, we can just perform a general testing of LMS and Studio using the following sandbox.

**Sandbox LMS:** https://pr31384.sandbox.opencraft.hosting/
**Sandbox Studio:** https://studio.pr31384.sandbox.opencraft.hosting/

## Deadline

"None" 

## Other information
Related PR: https://github.com/mitodl/edx-sga/pull/332 (it's not a blocker, as these integration tests do not seem to be used anywhere).

tecoholic committed:
refactor: rename ItemFactory to BlockFactory

openedx-webhooks commented:
Thanks for the pull request, @tecoholic!


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

tecoholic committed:
refactor: rename get_displayable_items and displayable_items

tecoholic commented:
@Agrendalath I was able to replace `item` the `ItemFactory` and the `get_display_items`  and `displayable_items` of the `XModuleMixin`. Other uses of `item` like [`load_item`](https://github.com/openedx/edx-platform/blob/master/xmodule/x_module.py#L1091) seem to have links to the [`modulestore`](https://github.com/openedx/edx-platform/blob/master/xmodule/modulestore/split_mongo/split.py#L771). Should we tackle these? 

Other heavy use of `item` is from the [`modulestore functions`](https://github.com/openedx/edx-platform/blob/master/xmodule/modulestore/__init__.py#L774) themselves. I guess, we are not touching these.

Agrendalath commented:
@tecoholic,

> Other uses of item like [load_item](https://github.com/openedx/edx-platform/blob/master/xmodule/x_module.py#L1091) seem to have links to the [modulestore](https://github.com/openedx/edx-platform/blob/master/xmodule/modulestore/split_mongo/split.py#L771). Should we tackle these?

We already have the [get_block](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/x_module.py#L1101-L1103) which uses `self.load_item`. We could replace all uses of `*.load_item` with `runtime.get_block` (or `self.get_block` in a few cases) and change `self.load_item` in `DescriptorSystem` to `self._load_item`, to avoid using it directly.

Regarding the name of the `load_item` argument of the `DescriptorSystem`: it makes sense to keep this name in sync with the methods passed to it. However, according to [this TODO](https://github.com/openedx/edx-platform/blob/4f5279d21418598dcab739b5dfd0364f0cd9f9ad/xmodule/modulestore/xml.py#L220-L222), it might be a good idea to rename these methods to `get_instance`. I'm specifically talking about the following occurrences:
1. [xmodule.x_module.DescriptorSystem](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/x_module.py#L1091-L1091)
2. [xmodule.modulestore.mongo.base:CachingDescriptorSystem](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/modulestore/mongo/base.py#L215-L305)
3. [xmodule.modulestore.split_mongo.split:SplitMongoModuleStore](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/modulestore/split_mongo/caching_descriptor_system.py#L86-L132)
4. [xmodule.modulestore.xml:ImportSystem](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/modulestore/xml.py#L220-L225)
5. [xmodule.tests.xml.\_\_init\_\_:InMemorySystem](https://github.com/openedx/edx-platform/blob/ee35ae6a5971d53cdfcf3f6e145a705ff2af5288/xmodule/tests/xml/__init__.py#L55-L58)

> Other heavy use of item is from the [modulestore functions](https://github.com/openedx/edx-platform/blob/master/xmodule/modulestore/__init__.py#L774) themselves. I guess, we are not touching these.

There are external packages that use these methods. For example, ORA2 and LTI_Consumer use `get_item` directly. I don't think we can safely rename these without proper deprecation, so we can omit them.

@ormsbee, what do you think about these points?

ormsbee commented:
@Agrendalath: I don't really have an opinion on the `get_instance` thing. That TODO was from 2012, and I honestly wouldn't read too much into it. I think it's fine to keep thing in sync with the methods passed to it.

Absolutely agree that we cannot break compatibility by changing the modulestore functions. We can add new methods and have the old ones call into them with deprecation warnings, but I'm honestly not sure if the long term simplification is worth having both methods around for the same thing for one or two release cycles.

tecoholic committed:
refactor: replace xmodule.load_item calls with get_block

tecoholic commented:
@Agrendalath @ormsbee Thank you for your inputs. I have changes most of the `system/runtime.load_item` calls to `get_block`. But making the `load_item` private by renaming it to `_load_item` breaks the system due to:

1. CachingDescriptorSystem defining the `load_item` function with a different signature [by adding extra kwargs](https://github.com/openedx/edx-platform/blob/18df1f020e35f48ec8681978f4c0d74031c26603/xmodule/modulestore/split_mongo/caching_descriptor_system.py#L86). It's usage cannot be replaced with the `get_block` without changing its signature.
2. The private property overshadowing the CachingDescriptorSystem's private memeber - [Lint Error](https://github.com/openedx/edx-platform/actions/runs/3619717582/jobs/6101090094#step:8:10)

Venturing further into renaming the `load_item` argument and syncing all the referencing functions seems a bit unfeasible at the moment, given we can't fully get rid of the `load_item` calls. And also, there doesn't seem to be any references to `runtime.load_item` outside of the `xmodule`

```sh
$ ack "load_item\(" -l
xmodule/tests/test_conditional.py
xmodule/tests/xml/__init__.py
xmodule/modulestore/split_mongo/split.py
xmodule/modulestore/split_mongo/caching_descriptor_system.py
xmodule/modulestore/xml.py
xmodule/modulestore/mongo/base.py
xmodule/x_module.py
```

Agrendalath commented on xmodule/modulestore/tests/factories.py:
```diff
@@ -279,7 +279,7 @@ def _create(cls, target_class, **kwargs):  # lint-amnesty, pylint: disable=argum
 
 
-class ItemFactory(XModuleFactory):
+class BlockFactory(XModuleFactory):
```
This change breaks integration tests of the [Staff Graded Assignment XBlock](https://github.com/mitodl/edx-sga/blob/c8e0f63604652e9c9c636f8a042d6869cf601841/edx_sga/tests/integration_tests.py#L35). We should open a PR to change it there too.
These tests seem to run only manually (not as a part of the CI), so this is not a blocker.

tecoholic replied:
I have opened the PR for `edx-sga` at https://github.com/mitodl/edx-sga/pull/332.

Agrendalath commented on conf/locale/en/LC_MESSAGES/django.po:
```diff
@@ -47,7 +47,7 @@ msgstr ""
 "Content-Transfer-Encoding: 8bit\n"
 "Plural-Forms: nplurals=2; plural=(n != 1);\n"
-"Generated-By: Babel 2.8.0\n"
+"Generated-By: Babel 2.11.0\n"
```
We didn't change anything else in this file. Why do we want to change this?

Also, we don't need to change `conf/locale/en/LC_MESSAGES/djangojs.po`.

tecoholic replied:
@Agrendalath Ah.. These came up as merge conflicts when I rebase on the master branch and I think I might have selected the wrong lines during resolution. I will revert the changes in the translation files.

Agrendalath commented on openedx/core/djangoapps/xblock/runtime/shims.py:
```diff
@@ -395,16 +395,16 @@ def has_dynamic_children(self):
         return False
 
-    def get_display_items(self):
+    def get_display_blocks(self):
```
This is a shim, added for backward compatibility. Changing its name doesn't bring too much value. If we want to get rid of it, we should:
1. Replace all uses of `get_display_items` with `get_children`.
2. Change [`xmodule.library_content_module.LibraryContentBlock.student_view()`](https://github.com/openedx/edx-platform/blob/dd1367823d0b7fbef39511c729366ed44ea45866/xmodule/library_content_module.py#L402-L408) to use `child` instead of looping through `child.displayable_items()`. That's because the latter only wraps the block in a list, so we're always iterating on a list with a single item.
3. Remove:
   1. `openedx.core.djangoapps.xblock.runtime.shims.XBlockShim.get_display_items()`
   3. `openedx.core.djangoapps.xblock.runtime.shims.XBlockShim.displayable_items()`
   4. `xmodule.x_module.XModuleMixin.get_display_items()`
   5. `xmodule.x_module.XModuleMixin.displayable_items()`

I haven't found any references to these methods outside of `edx-platform`, so it should be a safe change.

Does it make sense, @ormsbee?

tecoholic replied:
@Agrendalath I have made these changes and updated the PR. However, the CI has been complaining of `ItemFactory` in ` lms/djangoapps/discussion/rest_api/tests/test_views.py` which doesn't actually exist. Not sure why this is happening.

tecoholic replied:
@Agrendalath Rebasing on to the latest master made this issue go away.

Agrendalath commented on xmodule/conditional_module.py:
```diff
@@ -326,7 +326,7 @@ def get_required_module_descriptors(self):
         for location in self.sources_list:
             try:
-                descriptor = self.system.load_item(location)
+                descriptor = self.system.get_block(location)
```
Nit: as we're changing these methods, we could replace `system` with `runtime`. `system` is only a backward compatibility alias.

Agrendalath commented on openedx/core/djangoapps/xblock/runtime/shims.py:
```diff
@@ -395,16 +395,16 @@ def has_dynamic_children(self):
+            blocks.extend(child.displayable_blocks())
+        return blocks
 
     def displayable_items(self):
```
We didn't rename this one. See [my comment above](https://github.com/openedx/edx-platform/pull/31384#discussion_r1041331387) first, though.

tecoholic committed:
refactor: rename system with runtime on conditional_block

tecoholic committed:
refactor: replace get_display_blocks with get_children

tecoholic committed:
refactor: replace displayable_blocks with child

This removes the "displayable_blocks" property and replaces all the
usages with the "child" property.

tecoholic committed:
refactor: remove get_displayable_items and displayable_items from XModule

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -606,19 +606,19 @@ def _student_or_public_view(self, context, prereq_met, prereq_meta_info, banner_
         _ = self.runtime.service(self, "i18n").ugettext
-        display_items = self.get_display_items()
-        self._update_position(context, len(display_items))
+        display_blocks = self.get_children()
```
Nit: can we easily replace all instances of `display_blocks` with `children` in this file?

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: verified that XBlocks work correctly (especially the `LibraryContentBlock`)
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Agrendalath commented:
@tecoholic, please rebase the PR and squash the commits into logical chunks.

@ormsbee, would you mind doing a sanity check on this PR? Unlike in https://github.com/openedx/edx-platform/pull/31113, we're not breaking any compatibility here.

The only external usage of the `ItemFactory` was in the `edx-sga` XBlock's integration tests that are not a part of any pipeline. We have already opened [a PR](https://github.com/mitodl/edx-sga/pull/332) to synchronize these names there.

ormsbee commented:
@Agrendalath: Looks sane to me. It's only changing test-related files, right? Nothing in the modulestore itself? Aside from the load_item -> get_block change, which is equivalent...?

ormsbee commented:
FWIW, this is another change that's worth broadcasting loudly, since it will likely break folks who are actively working on branches.

Agrendalath commented:
@ormsbee, we have additionally (as described [here](https://github.com/openedx/edx-platform/pull/31384#discussion_r1041331387)):
1. [Replaced](https://github.com/open-craft/edx-platform/commit/f6e7dcc0a65445b19461a439a06f42a3ff534c1f) iterations over `displayable_items` with `self`, as this method was returning `[self]`.
2. [Replaced](https://github.com/open-craft/edx-platform/commit/a0bc78c839b7410a7daf9995243291197ab0a8cf) `displayable_items` with `get_children`, as the result is identical.
3. [Removed](https://github.com/open-craft/edx-platform/commit/2daa8ad5cae167b7fb109bb7a03449b48ebda852) these two (already deprecated) methods, as they were not used outside of the `edx-platform`.

These changes do not affect the modulestore, though.

> FWIW, this is another change that's worth broadcasting loudly, since it will likely break folks who are actively working on branches.

Good point. I will post an announcement once we merge this.

tecoholic committed:
refactor: rename display_blocks to children in seq_block

tecoholic commented:
@Agrendalath Thanks for the review. I have squashed the commits and rebased it on the latest changes. This is ready for merging.

tecoholic commented:
@Agrendalath Gentle reminder about merging this before some conflict with the master branch is introduced.

Agrendalath commented:
@ormsbee, I'm just checking if you had a chance to see [my previous comment](https://github.com/openedx/edx-platform/pull/31384#issuecomment-1361916558) and if there is anything else to do before we schedule the merge.

ormsbee commented:
@Agrendalath: No, go ahead and schedule the merge, ty

openedx-webhooks commented:
@tecoholic 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31384&entry.1484655318=https://github.com/tecoholic&entry.752974735=2022-12-01+05:48&entry.1917517419=2023-01-23+14:04&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

Agrendalath commented:
Forum announcement for this change: https://discuss.openedx.org/t/refactoring-xblock-references/9194.
